(DOCSP-40814) [C++] Specialized Data Formats > BSON (#81)
mongodb/docs-cpp: source/data-formats/working-with-bson.txt

   :name: genre
   :values: reference

Overview
--------
In this guide, you can learn how to use the {+driver-short+} to store and interact
with **BSON data** by using the ``bsoncxx`` library.  

BSON Data Format 
----------------

MongoDB uses the **BSON**, or Binary JSON, data format to store documents and
make remote procedure calls. This data format includes all JSON data structure types

> and supports additional types, including dates, different size integers, ObjectIds,

Reviewer: changed to ``ObjectId`` values

and binary data. For a complete list of supported types, see the :manual:`BSON Types </reference/bson-types/>`
page in the {+mdb-server+} manual. 

BSON Values and Views
~~~~~~~~~~~~~~~~~~~~~

> Many {+driver-short+} methods take a BSON document as an argument.

Reviewer 2: S: 'accept' sounds better to my ear? maybe more precise

```suggestion
Many {+driver-short+} methods accept a BSON document as an argument.
```

> The ``bsoncxx`` library provides two types that represent an underlying

Reviewer 2: S:

```suggestion
The ``bsoncxx`` library provides two data types that you can use to represent an underlying
```

> The ``bsoncxx`` library provides two types that represent an underlying

Reviewer 2: Q: what's an 'underlying' bson document? just the data the variable points to?

Reviewer: yes, but took it out since I think just "a BSON document" is fine

BSON document: ``bsoncxx::document::value`` and ``bsoncxx:document::view``. 

> The ``document::value`` type represents a BSON document that

Reviewer 2: is it clearer to say that a `document:value` object represents the BSON document?

Reviewer: Yeah I think it makes sense to say objects when referring to instances of the type - changed!

> owns its underlying buffer of data. When you pass a ``document::value`` to a

Reviewer 2: ```suggestion
owns its underlying buffer of data. When you pass a ``document::value`` object to a
```

> owns its underlying buffer of data. When you pass a ``document::value`` to a

Reviewer: I assume so, I don't think it's necessary to describe

> {+driver-short+} method, the method receives a copy of the ``document::value``.

Reviewer 2: Q: does this mean a copy of the object the variable points to?

Reviewer: I think a copy of the underlying BSON document data

> When a ``document::value`` type goes out of scope, its underlying buffer is freed.

Reviewer 2: would it be the variable that goes out of scope? or the object? i'm not very familiar with C++

Reviewer: The variable, I'll make that clear

object to a list of key-value pairs. The list builder creates a document if the key-value list fulfills
the following requirements:

- The list has an even number of elements.
- Each key is a ``string`` type and each value is a ``bson_value::value`` type or
  implicitly convertible to one.

If the preceding requirements are not met, the list builder creates an array. 

This example uses the list builder to perform the following actions:

- Build a document 
- Build an array
- Convert the builder document to a ``bsoncxx::document::value``

.. literalinclude:: /includes/data-formats/bson.cpp
    :language: cpp
    :dedent:
    :start-after: start-bson-list
    :end-before: end-bson-list

.. _cpp-bson-basic:

Basic Builder
~~~~~~~~~~~~~

The ``builder::basic`` interface is a builder-style interface for constructing a BSON document. 
To create a BSON document by using the basic builder, specify your document's data in a list of
key-value pair objects. You can create these key-value pair objects by passing a key and value
to the ``builder::basic::kvp()`` method. The key must be a ``string`` type, and

> the value must be a ``bson_value::value`` type or implicitly convertible to one. 

Reviewer: it's really one step described here (creating the key-value pairs). Then, you can either pass them to the make_document() method, which is described in the next paragraph, or construct a builder::basic::document and append the key-value pairs, described a little further down

    :language: cpp
    :dedent:
    :start-after: start-bson-make-document
    :end-before: end-bson-make-document

Alternatively, you can create a BSON document across multiple statements
by appending key-value pairs to a basic builder object. This 
example uses the basic builder to perform the following actions:

- Initialize a ``builder::basic::document`` object
- Store data in the document by using the ``append()`` method
- Convert the builder document to a ``bsoncxx::document::value``

.. literalinclude:: /includes/data-formats/bson.cpp
    :language: cpp
    :dedent:
    :start-after: start-bson-append
    :end-before: end-bson-append

.. _cpp-bson-stream:

Stream Builder
~~~~~~~~~~~~~~

.. important::

   We recommend using the :ref:`Basic Builder <cpp-bson-basic>` instead 
   of the stream builder. 
   
   To properly append each new value, a stream builder must keep track
   of the state of the current document. You cannot reuse the initial
   stream builder after this state changes. As a result, all intermediate
   values must be stored in new variables if you build a document
   across multiple statements. Due to this complexity, use of the stream

>    builder is discouraged.

Reviewer: I'll leave this section for now and ask the tech reviewer if there are use cases or if I should just remove the section

Overview
--------
In this guide, you can learn how to use the {+driver-short+} to store and interact
with BSON data by using the ``bsoncxx`` library.  

BSON Data Format 
----------------

MongoDB uses the **BSON**, or Binary JSON, data format to store documents and
make remote procedure calls. This data format includes all JSON data structure types
and supports additional types, including dates, different-sized integers, ``ObjectId``
values, and binary data. For a complete list of supported types, see the :manual:`BSON Types </reference/bson-types/>`
page in the {+mdb-server+} manual. 

BSON Values and Views
~~~~~~~~~~~~~~~~~~~~~

Many {+driver-short+} methods accept a BSON document as an argument.
The ``bsoncxx`` library provides two data types that you can use to represent
a BSON document: ``bsoncxx::document::value`` and ``bsoncxx:document::view``. 

A ``document::value`` object represents a BSON document that owns its
underlying buffer of data. When you pass a ``document::value`` object
to a {+driver-short+} method, the method receives a copy of the BSON document
data. When a ``document::value`` object goes out of scope, its underlying buffer is freed.

.. tip::

>    You can convert builder types to ``document::value`` types by calling the ``extract()``
>    method on the builder. To learn more about the builder types, see the :ref:`cpp-bson-builders`

Reviewer 3: ```suggestion
   You can convert builder types to ``document::value`` types by calling methods on the builder. To learn more about the builder types, see the :ref:`cpp-bson-builders`
```

`bsoncxx::builder::list` does not have an `extract` method.

key-value pair objects. You can create these key-value pair objects by passing a key and value
to the ``builder::basic::kvp()`` method. The key must be a ``string`` type, and
the value must be a ``bson_value::value`` type or implicitly convertible to one. 

You can use the basic builder's ``make_document()`` method to create a document and
convert it to a ``bsoncxx::document::value`` in a single statement, as shown in the
following code:

.. literalinclude:: /includes/data-formats/bson.cpp
    :language: cpp
    :dedent:
    :start-after: start-bson-make-document
    :end-before: end-bson-make-document

Alternatively, you can create a BSON document across multiple statements
by appending key-value pairs to a basic builder object. This 
example uses the basic builder to perform the following actions:

- Initialize a ``builder::basic::document`` object
- Store data in the document by using the ``append()`` method
- Convert the builder document to a ``bsoncxx::document::value`` object

.. literalinclude:: /includes/data-formats/bson.cpp
    :language: cpp
    :dedent:
    :start-after: start-bson-append
    :end-before: end-bson-append

.. _cpp-bson-stream:

> Stream Builder

Reviewer: Note to tech reviewer: should we document the stream builder or just remove this section?

Reviewer 3: Sorry, I missed this on first review. I support keeping the documentation for the stream builder with the warning to discourage use. The stream builder is currently [documented](https://github.com/mongodb/mongo-cxx-driver/blob/7402d3d8ffc093fdc647fb2b2469f3a154beb41e/src/bsoncxx/include/bsoncxx/v_noabi/bsoncxx/builder/stream/document.hpp#L35) as discouraged, but not-yet deprecated or planned for removal.

~~~~~~~~~~~~

The ``builder::list`` interface is a JSON-like builder for constructing documents and arrays. 
To create a BSON document by using the list builder, construct a ``bsoncxx::builder::list``
object, passing a list of key-value pairs to the constructor. The list builder creates a
BSON document if the key-value list fulfills the following requirements:

- The list has an even number of elements.

> - Each key is a ``string`` type and each value is a ``bson_value::value`` type is or

Reviewer 2: ```suggestion
- Each key is a ``string`` type and each value is a ``bson_value::value`` type or is
```
